cpp/src/arrow/compute/kernels/vector_selection_filter_internal.cc

```diff
@@ -1039,10 +1039,17 @@ class FilterMetaFunction : public MetaFunction {
 
     if (args[0].kind() == Datum::RECORD_BATCH) {
       auto values_batch = args[0].record_batch();
-      ARROW_ASSIGN_OR_RAISE(
-          std::shared_ptr<RecordBatch> out_batch,
-          FilterRecordBatch(*args[0].record_batch(), args[1], options, ctx));
-      return Datum(out_batch);
+      if (args[1].kind() == Datum::ARRAY) {
+        ARROW_ASSIGN_OR_RAISE(std::shared_ptr<RecordBatch> out_batch,
+                              FilterRecordBatch(*values_batch, args[1], options, ctx));
+        return Datum(out_batch);
+      } else {
```

Comment:
Wouldn't it make sense to change `FilterRecordBatch` to handle ChunkedArrays instead?

The problem seems to be that `FilterRecordBatch` takes for granted that the filter is always an array ( https://github.com/apache/arrow/blob/6e1b62509b5e0ba69cad630519e3757688a1e0c1/cpp/src/arrow/compute/kernels/vector_selection_filter_internal.cc#L931 )

But it shouldn't be too hard to check the filter type and handle it like what `FilterTable` does https://github.com/apache/arrow/blob/6e1b62509b5e0ba69cad630519e3757688a1e0c1/cpp/src/arrow/compute/kernels/vector_selection_filter_internal.cc#L962-L971

in the end you could write the code to always work with multiple chunks and consider a single array a case where there is just one chunk

Reply:
`FilterRecordBatch` is only used for `FilterMetaFunction::ExecuteImpl`, so it's not that expanding the capabilities is useful in general for other use cases. So just for the kernel exec, to me it seems like unnecessary duplication to expand `FilterRecordBatch` to also rechunk the input, given that the conversion of RecordBatch to Table is a cheap zero-copy one.

Reply:
Additionally, this keeps the return value of filtering a RecordBatch with Array as RecordBatch, and not Table, which it would be it `FilterRecordBatch` gets generalized (unless we add logic to convert back the Table to RecordBatch in that case)

Reply:
Agreeing with Joris on this.

Reply:
I'm not sure why we would want a filtering a RecordBatch to return a Table, outcome of filtering will never be chunked even when the filter itself was chunked.

Also, given that the filter is generally <= to the filtered array length, we can probably take for granted that it will always be possible to concatenate the filter when it's a chunked array. I think it can be benchmarked, but if I had to guess I'd say that concatenating the chunks of the filter has a good chance to be faster than FilterTable

Reply:
> I'm not sure why we would want a filtering a RecordBatch to return a Table, outcome of filtering a RecordBatch will never be chunked even when the filter itself was chunked.

I was only saying that because that is what would happen if we use the same approach as `FilterTable` does (which I thought you were suggesting, but rereading you only say to check the filter type like FilterTable does)

But it's a good point that we could also just concatenate the filter. Type stability/predictability wise, it's probably nice to have RecordBatch.filter() always return a RecordBatch.